File: lib/schema-validator.ts
/**
 * JSON schema validation for data files (time logs, costs, revenue).
 */

export interface TimeEntry {
  timestamp: string;
  activity: string;
  client_id?: string | null;
  duration_seconds: number;
  time_saved_seconds?: number;
  metadata?: Record<string, any>;
}

export interface ApiCostEntry {
  timestamp: string;
  provider: string;
  model: string;
  activity: string;
  client_id?: string | null;
  input_tokens: number;
  output_tokens: number;
  cost_usd: number;
  metadata?: Record<string, any>;
}

export interface HostingCostEntry {
  timestamp: string;
  client_id: string;
  cost_usd: number;
  type: "hosting";
}

export interface RevenueEntry {
  timestamp: string;
  client_id?: string | null;
  type: string;
  amount_usd: number;
  package?: string;
}

/**
 * Validate a time log entry.
 */
export function validateTimeEntry(entry: any): { valid: boolean; error?: string } {
  if (!entry || typeof entry !== "object") {
    return { valid: false, error: "Entry must be an object" };
  }

  if (typeof entry.timestamp !== "string") {
    return { valid: false, error: "timestamp must be a string" };
  }

  if (typeof entry.activity !== "string") {
    return { valid: false, error: "activity must be a string" };
  }

  if (entry.client_id !== null && entry.client_id !== undefined && typeof entry.client_id !== "string") {
    return { valid: false, error: "client_id must be a string or null" };
  }

  if (typeof entry.duration_seconds !== "number") {
    return { valid: false, error: "duration_seconds must be a number" };
  }

  if (entry.time_saved_seconds !== undefined && typeof entry.time_saved_seconds !== "number") {
    return { valid: false, error: "time_saved_seconds must be a number" };
  }

  if (entry.metadata !== undefined && (typeof entry.metadata !== "object" || Array.isArray(entry.metadata))) {
    return { valid: false, error: "metadata must be an object" };
  }

  return { valid: true };
}

/**
 * Validate an API cost entry.
 */
export function validateApiCostEntry(entry: any): { valid: boolean; error?: string } {
  if (!entry || typeof entry !== "object") {
    return { valid: false, error: "Entry must be an object" };
  }

  if (typeof entry.timestamp !== "string") {
    return { valid: false, error: "timestamp must be a string" };
  }

  if (typeof entry.provider !== "string") {
    return { valid: false, error: "provider must be a string" };
  }

  if (typeof entry.model !== "string") {
    return { valid: false, error: "model must be a string" };
  }

  if (typeof entry.activity !== "string") {
    return { valid: false, error: "activity must be a string" };
  }

  if (entry.client_id !== null && entry.client_id !== undefined && typeof entry.client_id !== "string") {
    return { valid: false, error: "client_id must be a string or null" };
  }

  if (typeof entry.input_tokens !== "number") {
    return { valid: false, error: "input_tokens must be a number" };
  }

  if (typeof entry.output_tokens !== "number") {
    return { valid: false, error: "output_tokens must be a number" };
  }

  if (typeof entry.cost_usd !== "number") {
    return { valid: false, error: "cost_usd must be a number" };
  }

  if (entry.metadata !== undefined && (typeof entry.metadata !== "object" || Array.isArray(entry.metadata))) {
    return { valid: false, error: "metadata must be an object" };
  }

  return { valid: true };
}

/**
 * Validate a hosting cost entry.
 */
export function validateHostingCostEntry(entry: any): { valid: boolean; error?: string } {
  if (!entry || typeof entry !== "object") {
    return { valid: false, error: "Entry must be an object" };
  }

  if (typeof entry.timestamp !== "string") {
    return { valid: false, error: "timestamp must be a string" };
  }

  if (typeof entry.client_id !== "string") {
    return { valid: false, error: "client_id must be a string" };
  }

  if (typeof entry.cost_usd !== "number") {
    return { valid: false, error: "cost_usd must be a number" };
  }

  if (entry.type !== "hosting") {
    return { valid: false, error: "type must be 'hosting'" };
  }

  return { valid: true };
}

/**
 * Validate a revenue entry.
 */
export function validateRevenueEntry(entry: any): { valid: boolean; error?: string } {
  if (!entry || typeof entry !== "object") {
    return { valid: false, error: "Entry must be an object" };
  }

  if (typeof entry.timestamp !== "string") {
    return { valid: false, error: "timestamp must be a string" };
  }

  if (entry.client_id !== null && entry.client_id !== undefined && typeof entry.client_id !== "string") {
    return { valid: false, error: "client_id must be a string or null" };
  }

  if (typeof entry.type !== "string") {
    return { valid: false, error: "type must be a string" };
  }

  if (typeof entry.amount_usd !== "number") {
    return { valid: false, error: "amount_usd must be a number" };
  }

  if (entry.package !== undefined && typeof entry.package !== "string") {
    return { valid: false, error: "package must be a string" };
  }

  return { valid: true };
}

/**
 * Validate an array of time log entries.
 */
export function validateTimeLogs(data: any[]): { valid: boolean; errors: string[] } {
  if (!Array.isArray(data)) {
    return { valid: false, errors: ["Time logs must be an array"] };
  }

  const errors: string[] = [];
  data.forEach((entry, index) => {
    const result = validateTimeEntry(entry);
    if (!result.valid) {
      errors.push(`Entry ${index}: ${result.error}`);
    }
  });

  return { valid: errors.length === 0, errors };
}

/**
 * Validate an array of cost entries (API or hosting).
 */
export function validateCostEntries(
  data: any[],
  entryType: "api" | "hosting"
): { valid: boolean; errors: string[] } {
  if (!Array.isArray(data)) {
    return { valid: false, errors: [`${entryType} costs must be an array`] };
  }

  const errors: string[] = [];
  const validator =
    entryType === "api" ? validateApiCostEntry : validateHostingCostEntry;

  data.forEach((entry, index) => {
    const result = validator(entry);
    if (!result.valid) {
      errors.push(`Entry ${index}: ${result.error}`);
    }
  });

  return { valid: errors.length === 0, errors };
}

/**
 * Validate an array of revenue entries.
 */
export function validateRevenueEntries(data: any[]): { valid: boolean; errors: string[] } {
  if (!Array.isArray(data)) {
    return { valid: false, errors: ["Revenue entries must be an array"] };
  }

  const errors: string[] = [];
  data.forEach((entry, index) => {
    const result = validateRevenueEntry(entry);
    if (!result.valid) {
      errors.push(`Entry ${index}: ${result.error}`);
    }
  });

  return { valid: errors.length === 0, errors };
}

File: automation/schema_validator.py
"""
JSON schema validation for data files (time logs, costs, revenue).

Provides validation functions to ensure data files match expected schemas.
"""
import logging
from typing import List, Dict, Any, Optional


def validate_time_entry(entry: Dict[str, Any]) -> tuple[bool, Optional[str]]:
    """
    Validate a time log entry against schema.
    
    Expected schema:
    {
        "timestamp": str (ISO format),
        "activity": str,
        "client_id": str | null,
        "duration_seconds": number,
        "time_saved_seconds": number,
        "metadata": dict (optional)
    }
    
    Returns:
        (is_valid, error_message)
    """
    if not isinstance(entry, dict):
        return False, "Entry must be a dictionary"
    
    # Required fields
    required = ["timestamp", "activity", "duration_seconds", "time_saved_seconds"]
    for field in required:
        if field not in entry:
            return False, f"Missing required field: {field}"
    
    # Validate types
    if not isinstance(entry["timestamp"], str):
        return False, "timestamp must be a string"
    
    if not isinstance(entry["activity"], str):
        return False, "activity must be a string"
    
    if entry["client_id"] is not None and not isinstance(entry["client_id"], str):
        return False, "client_id must be a string or null"
    
    if not isinstance(entry["duration_seconds"], (int, float)):
        return False, "duration_seconds must be a number"
    
    if not isinstance(entry["time_saved_seconds"], (int, float)):
        return False, "time_saved_seconds must be a number"
    
    # Optional metadata
    if "metadata" in entry and not isinstance(entry["metadata"], dict):
        return False, "metadata must be a dictionary"
    
    return True, None


def validate_api_cost_entry(entry: Dict[str, Any]) -> tuple[bool, Optional[str]]:
    """
    Validate an API cost entry against schema.
    
    Expected schema:
    {
        "timestamp": str (ISO format),
        "provider": str,
        "model": str,
        "activity": str,
        "client_id": str | null,
        "input_tokens": number,
        "output_tokens": number,
        "cost_usd": number,
        "metadata": dict (optional)
    }
    
    Returns:
        (is_valid, error_message)
    """
    if not isinstance(entry, dict):
        return False, "Entry must be a dictionary"
    
    # Required fields
    required = ["timestamp", "provider", "model", "activity", "input_tokens", "output_tokens", "cost_usd"]
    for field in required:
        if field not in entry:
            return False, f"Missing required field: {field}"
    
    # Validate types
    if not isinstance(entry["timestamp"], str):
        return False, "timestamp must be a string"
    
    if not isinstance(entry["provider"], str):
        return False, "provider must be a string"
    
    if not isinstance(entry["model"], str):
        return False, "model must be a string"
    
    if not isinstance(entry["activity"], str):
        return False, "activity must be a string"
    
    if entry["client_id"] is not None and not isinstance(entry["client_id"], str):
        return False, "client_id must be a string or null"
    
    if not isinstance(entry["input_tokens"], int):
        return False, "input_tokens must be an integer"
    
    if not isinstance(entry["output_tokens"], int):
        return False, "output_tokens must be an integer"
    
    if not isinstance(entry["cost_usd"], (int, float)):
        return False, "cost_usd must be a number"
    
    # Optional metadata
    if "metadata" in entry and not isinstance(entry["metadata"], dict):
        return False, "metadata must be a dictionary"
    
    return True, None


def validate_hosting_cost_entry(entry: Dict[str, Any]) -> tuple[bool, Optional[str]]:
    """
    Validate a hosting cost entry against schema.
    
    Expected schema:
    {
        "timestamp": str (ISO format),
        "client_id": str,
        "cost_usd": number,
        "type": "hosting"
    }
    
    Returns:
        (is_valid, error_message)
    """
    if not isinstance(entry, dict):
        return False, "Entry must be a dictionary"
    
    # Required fields
    required = ["timestamp", "client_id", "cost_usd", "type"]
    for field in required:
        if field not in entry:
            return False, f"Missing required field: {field}"
    
    # Validate types
    if not isinstance(entry["timestamp"], str):
        return False, "timestamp must be a string"
    
    if not isinstance(entry["client_id"], str):
        return False, "client_id must be a string"
    
    if not isinstance(entry["cost_usd"], (int, float)):
        return False, "cost_usd must be a number"
    
    if entry["type"] != "hosting":
        return False, "type must be 'hosting'"
    
    return True, None


def validate_revenue_entry(entry: Dict[str, Any]) -> tuple[bool, Optional[str]]:
    """
    Validate a revenue entry against schema.
    
    Expected schema:
    {
        "timestamp": str (ISO format),
        "client_id": str | null,
        "type": str (e.g., "deposit", "final_payment", "hosting"),
        "amount_usd": number,
        "package": str (optional)
    }
    
    Returns:
        (is_valid, error_message)
    """
    if not isinstance(entry, dict):
        return False, "Entry must be a dictionary"
    
    # Required fields
    required = ["timestamp", "type", "amount_usd"]
    for field in required:
        if field not in entry:
            return False, f"Missing required field: {field}"
    
    # Validate types
    if not isinstance(entry["timestamp"], str):
        return False, "timestamp must be a string"
    
    if entry["client_id"] is not None and not isinstance(entry["client_id"], str):
        return False, "client_id must be a string or null"
    
    if not isinstance(entry["type"], str):
        return False, "type must be a string"
    
    if not isinstance(entry["amount_usd"], (int, float)):
        return False, "amount_usd must be a number"
    
    # Optional package field
    if "package" in entry and not isinstance(entry["package"], str):
        return False, "package must be a string"
    
    return True, None


def validate_time_logs(data: List[Dict[str, Any]]) -> tuple[bool, Optional[str], int]:
    """
    Validate an array of time log entries.
    
    Returns:
        (is_valid, error_message, invalid_count)
    """
    if not isinstance(data, list):
        return False, "Time logs must be an array", 0
    
    invalid_count = 0
    for i, entry in enumerate(data):
        is_valid, error = validate_time_entry(entry)
        if not is_valid:
            invalid_count += 1
            logging.warning(f"Invalid time entry at index {i}: {error}")
    
    if invalid_count > 0:
        return False, f"{invalid_count} invalid entries found", invalid_count
    
    return True, None, 0


def validate_cost_entries(data: List[Dict[str, Any]], entry_type: str = "api") -> tuple[bool, Optional[str], int]:
    """
    Validate an array of cost entries (API or hosting).
    
    Args:
        data: Array of cost entries
        entry_type: "api" or "hosting"
    
    Returns:
        (is_valid, error_message, invalid_count)
    """
    if not isinstance(data, list):
        return False, f"{entry_type} costs must be an array", 0
    
    invalid_count = 0
    validator = validate_api_cost_entry if entry_type == "api" else validate_hosting_cost_entry
    
    for i, entry in enumerate(data):
        is_valid, error = validator(entry)
        if not is_valid:
            invalid_count += 1
            logging.warning(f"Invalid {entry_type} cost entry at index {i}: {error}")
    
    if invalid_count > 0:
        return False, f"{invalid_count} invalid entries found", invalid_count
    
    return True, None, 0


def validate_revenue_entries(data: List[Dict[str, Any]]) -> tuple[bool, Optional[str], int]:
    """
    Validate an array of revenue entries.
    
    Returns:
        (is_valid, error_message, invalid_count)
    """
    if not isinstance(data, list):
        return False, "Revenue entries must be an array", 0
    
    invalid_count = 0
    for i, entry in enumerate(data):
        is_valid, error = validate_revenue_entry(entry)
        if not is_valid:
            invalid_count += 1
            logging.warning(f"Invalid revenue entry at index {i}: {error}")
    
    if invalid_count > 0:
        return False, f"{invalid_count} invalid entries found", invalid_count
    
    return True, None, 0

